vmm/src/memory_manager.rs

```diff
@@ -450,14 +505,34 @@ impl MemoryManager {
         Ok(slot)
     }
 
-    pub fn resize(&mut self, desired_ram: u64) -> Result<bool, Error> {
-        if desired_ram > self.current_ram {
-            self.hotplug_ram_region((desired_ram - self.current_ram) as usize)?;
-            self.current_ram = desired_ram;
-            Ok(true)
+    pub fn virtiomem_resize(&mut self, size: u64) -> Result<(), Error> {
+        let region = self.virtiomem_region.take();
+        if let Some(region) = region {
+            self.add_region(region)?;
+        }
+
+        if let Some(resize) = &self.virtiomem_resize {
+            resize.work(size).map_err(Error::VirtioMemResizeFail)?;
         } else {
-            Ok(false)
+            panic!("should not fail here");
+        }
+
+        Ok(())
+    }
+
+    pub fn resize(&mut self, desired_ram: u64) -> Result<bool, Error> {
+        let mut ret = false;
+        if desired_ram >= self.boot_ram {
+            if self.virtiomem_resize.is_some() {
+                self.virtiomem_resize(desired_ram - self.boot_ram)?;
+                self.current_ram = desired_ram;
+            } else if self.current_ram > desired_ram {
```

Comment:
I think you meant `desired_ram < self.current_ram` here, otherwise we end up calling `hotplug_ram_region()` with negative size.

Overall, I think if you implement something similar to what I suggested in https://github.com/cloud-hypervisor/cloud-hypervisor/pull/837#discussion_r394185636, then this routine would be more readable as follows:

```Rust
    pub fn resize(&mut self, desired_ram: u64) -> Result<bool, Error> {
        match self.hotplug_method {
            HotplugMethod::VirtioMem => {
                if desired_ram >= self.boot_ram {
                    self.virtiomem_resize(desired_ram - self.boot_ram)?;
                    self.current_ram = desired_ram;
                }
            }
            HotplugMethod::Acpi => {
                if desired_ram >= self.current_ram {
                    self.hotplug_ram_region((desired_ram - self.current_ram) as usize)?;
                    self.current_ram = desired_ram;
                } else {
                    return Ok(false);
                }
            }
        }

        Ok(true)
    }  
```

Reply:
Change this part to:
```
    pub fn resize(&mut self, desired_ram: u64) -> Result<bool, Error> {
        let mut notify_hotplug = false;
        match self.hotplug_method {
            HotplugMethod::VirtioMem => {
                if desired_ram >= self.boot_ram {
                    self.virtiomem_resize(desired_ram - self.boot_ram)?;
                    self.current_ram = desired_ram;
                }
            }
            HotplugMethod::Acpi => {
                if desired_ram >= self.current_ram {
                    self.hotplug_ram_region((desired_ram - self.current_ram) as usize)?;
                    self.current_ram = desired_ram;
                    notify_hotplug = true
                }
            }
        }
        Ok(notify_hotplug)
    }
```